### PR TITLE
plumbing: format/index perf, buffered reads, reflection removal

### PR DESCRIPTION
**Pull request by saracen**
Large performance increase by buffering reads.

There were a few instances where binary.Read() would end up using reflection on &plumbing.Hash, rather than treating it as a byte slice. This has now been resolved.

### CODE REVIEW

File: utils/binary/read.go [hunk 1]
```diff
@@ -25,6 +25,15 @@ func Read(r io.Reader, data ...interface{}) error {
 
 // ReadUntil reads from r untin delim is found
 func ReadUntil(r io.Reader, delim byte) ([]byte, error) {
+	if bufr, ok := r.(*bufio.Reader); ok {
```
Comment (mcuadros) [inline utils/binary/read.go]: Can you move this to a function called `ReadUntilFromBufioReader` or something like that, and call it from here?

Comment (saracen) [reply utils/binary/read.go]: Yeah, will push shortly with a test for it.

The only problem with this approach is that bufio.Reader satisfies io.Reader too, so if you pass a bufio.Reader to it, it's not going to be immediately obvious you're getting slower behaviour.

There's very few places using ReadUntil, so it's not an issue now. But it's something to look out for in the future and to ensure that if the reader is buffered, they're using the correct variant.

Comment (mcuadros) [reply utils/binary/read.go]: I mean call one from the other, while the new behavior is public

Comment (mcuadros) [reply utils/binary/read.go]:
```
if bufr, ok := r.(*bufio.Reader); ok {
	return ReadUntilFromBufioReader(bufr)
}
```

Comment (saracen) [reply utils/binary/read.go]: Done! Sorry about that, re-reading your comment, it was clear enough. My bad.

**Pull request merged**